src/librustc_trans/type_of.rs

```diff
@@ -9,231 +9,472 @@
 // except according to those terms.
 
 use abi::FnType;
-use adt;
 use common::*;
-use machine;
+use rustc::hir;
 use rustc::ty::{self, Ty, TypeFoldable};
-use rustc::ty::layout::LayoutTyper;
+use rustc::ty::layout::{self, Align, LayoutOf, Size, TyLayout};
+use rustc_back::PanicStrategy;
 use trans_item::DefPathBasedNames;
 use type_::Type;
 
-use syntax::ast;
+use std::fmt::Write;
 
-pub fn fat_ptr_base_ty<'a, 'tcx>(ccx: &CrateContext<'a, 'tcx>, ty: Ty<'tcx>) -> Type {
-    match ty.sty {
-        ty::TyRef(_, ty::TypeAndMut { ty: t, .. }) |
-        ty::TyRawPtr(ty::TypeAndMut { ty: t, .. }) if !ccx.shared().type_is_sized(t) => {
-            in_memory_type_of(ccx, t).ptr_to()
+fn uncached_llvm_type<'a, 'tcx>(ccx: &CrateContext<'a, 'tcx>,
+                                layout: TyLayout<'tcx>,
+                                defer: &mut Option<(Type, TyLayout<'tcx>)>)
+                                -> Type {
+    match layout.abi {
+        layout::Abi::Scalar(_) => bug!("handled elsewhere"),
+        layout::Abi::Vector => {
+            return Type::vector(&layout.field(ccx, 0).llvm_type(ccx),
+                                layout.fields.count() as u64);
         }
-        ty::TyAdt(def, _) if def.is_box() => {
-            in_memory_type_of(ccx, ty.boxed_ty()).ptr_to()
+        layout::Abi::ScalarPair(..) => {
```

Comment:
It is `{ i1, i1 }`. `i1` occupies 1 byte in memory. We don't use it for memory accesses because LLVM is just bad. There is no fundamental reason either way.

Reply:
Could you at least document this somewhere (the thing where we use `i1` in struct decls and it's OK but we must use `i8*` for loading to avoid problems), preferably with a (codegen) test?

Reply:
documented in `scalar_pair_element_llvm_type`